---
title: "Other notes from Bioinformatics Summer School 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
```

# Day 1 

## Morgan, intro R

```{r}
getwd()
```


`methods` lists all methods available for a given class, but not simple functions:

```{r}
methods(class = lm)
```
Getting specific help for a method call:
```{r, eval = FALSE}
?plot.lm
```

Note that not all functions are shown e.g. abline, which is a pure function:

```{r}
abline
```
where e.g. `anova` is a generic that dispatches to various methods:
```{r}
anova
```

and methods is actually listing the generics that operate. 
 
## Bioconductor intro

`nchar` counts the number of characters in each entry of character vector:

```{r}
seq = c("AAACA", "CATGC")
nchar(seq)
```

```{r}
"DNAStringSet"
```

Check available packages (not case sensitive for search):

```{r}
library(BiocManager)
BiocManager::available("readr")
```


## classes, methods etc



```{r}
class(anova); anova
```

anova is an S3 class (all things such as generic.method are S3)

```{r}
fit<- lm(Sepal.Length~Sepal.Width , iris)
anova
```


```{r}
getAnywhere(anova.lm)
```
```{r}
library(Biostrings)
```
```{r, eval= FALSE}
?`reverseComplement,DNAStringSet-class`  
```

This structure is for S4:
```{r, eval = FALSE}
?`reverseComplement,DNAStringSet-method`
```


# Day 2 

## Bioconductor core technologies - Morgan

### Genomic ranges

Map reads to known ranges in reference genome, identify gene and then determine expression by couting the number of matches. 

Variants have known genomic coordinates too i.e. a genomic range. Example question: Do they affect expression? First match to gene, then locate it in genome and identify what it may do. 

```{r}
library(GenomicRanges)
exon = GRanges(c("chr1:20-30:+","chr1:40-50:+","chr1:45-55:+"))
exon
```
Recall, most protein-coding genes in eukaryotic cells are composed
of a number of coding regions (exons) interspersed with
noncoding regions (introns) 

(see [Documentation/BioInformatic/0.0_DNA_and_the_flow_of_genetic_information_2015.pdf])

```{r}
start(exon)
```
```{r}
end(exon)
```
```{r}
width(exon)
```
Note there are different ways to count: half-open or closed. Representation assumes it is closed intervals.
```{r}
snp<- GRanges(c("chr:12345","chr:23456"))
```

There is a shift operator
```{r}
shift(snp,1)
```

and flank:
```{r}
flank(snp,10)
```
```{r}
reduce(exon)
```

```{r}
disjoin(exon)
```
```{r}
snp<-GRanges(c("chr1:25","chr1:35","chr1:48"))
snp
```

```{r}
exon
```
The following does the obvious count: how many times the snips overlap each exon

```{r}
countOverlaps(snp,exon)
```

```{r}
countOverlaps(exon,snp)
```

again doing the obvious: not symmetric. 


```{r}
snp %over% exon
```
```{r}
snp[snp %over% exon]
```
```{r}
cvg<- coverage(exon)
cvg
```
We read this as follows: "Start a position 1 there are 19 zeros, then then are 11 with 1 exon, then 9 zeros, then 5 1s, 6 2s etc." 

You can also coerce this back to a GRanges object:

```{r}
(grcvg<- cvg %>% as("GRanges"))
```

but you lose the strange information.

You can pull out the metadata, which is on the rhs of the | in the GRanges object. 

```{r}
mcols(grcvg)
```

Note that this is a different kind of object: a DataFrame (capital D).

```{r}
library(Biostrings)
df<- DataFrame(
  i=1:3,
  dna = DNAStringSet(c("AAA","CCC","GGG")),
  gr = exon
  
  
)
df
```

## Data visualisation

[Modern Statistics for Modern Biology, by Susan Holmes, Wolfgang Huber](http://web.stanford.edu/class/bios221/book/)



# Day 3

# Gene Set Enrichment Analysis

### Approaches (with practical) 

See [Kharti et al 2012 ](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375)

A *gene set* is any a priori classification of 'genes' into biological relevant groups. 

The gene ontology is a DAG. 

Approach 3: no single gene is significantly different, but the entire pathway is over-expressed in one group. Test on aggregrate of single gene tests. Permutations. [Goemann and Buehlmann 2007](https://www.ncbi.nlm.nih.gov/pubmed/17303618).

Approach 4: Linear model approach [Hummel et al 2008](https://academic.oup.com/bioinformatics/article/24/1/78/205159)

The above appraoches were mainly developed on microarray data. The following were developed for sequence data. 

Aprroach 5: issues
All beign equal, long genes receive more reads than short genes. See [Young et al](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-2-r14). Normalise by gene length. 

Aprroach 6: de novo discovery. Consider the previous as being supervised. Now try unsupervised. [Langfelder and Horvath](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559)

### Benchmarks

Levi Waldron: [GSEABenchmarkeR](https://github.com/waldronlab/GSEABenchmarkeR) and the twitter description [here](https://twitter.com/LeviWaldron1/status/1142092301403115521)

### Issues with single cell data

Dim reduction, therefore should analysis weightings? 


### Slides wrap up

[BiocView](https://www.bioconductor.org/packages/devel/BiocViews.html)

### Rmd 6

Volcano plots - size v. stat significance 

goana: does the relevant joins to get the gene sets from the entrez ids. 


## Intro Stats and ML - Oliver Crook

p-value histogram is very informative/ should be uniform, spikes at higher values suggest the wrong test, small slope upwards towards wero implies lower power.

# Day 4

## High-throughput sequencing data processing: Charlotte Sonenson

mRNA code for protein, need to remove ribosome RNA.

Fragments are around 300 nucleotides. Don't know which mRNA strand is original or compliment. Can label original strand. Can check by mapping strands to genome. B on slide 6 is indicative of stranded protocol. 

Hiseq illumina: 100-200 nucleotides. 

Record phred score to letter, adding 33 t ophred as first 33 ascii characters are invisible. Captitals are therefore showing better quality reads; 

### Alignment based approach: Count reads per gene. 

Align reads to genes.recall reads are from transcripts (don't know if splice etc was there in genome). One read can align to multuple genes

Split reads across introns

Take all reads that overlaps an exon and sum them up. 

Long transripts = more reads, therefore larger sum. GC content affects number of reads

Can count reads per gene or reads per exon with a gene. Can get reads per exon. Can check if same or different isoforms expressed. Is one exon missing v. present in different conditions. 

Get annotations from public sources. 



Online data: typical want primary assembly. DNA: genome. cDNA: coing DNA - transcripts.

Need to know where genes are, obviously: gtf file; One row for each genomic feature. e.g. 2R = chromosome. GTF files in gene set column. 


Many aligners exist. For RNA seq, the alinger needs to aware of the splicer. STAR more established. HISAT2 uses less memory. 

**Step 1 - index genome**

Compare each read to each base in the genome (3 billion base pairs for humans). Too slow, so make index of genome. Index genome once (genomeGenerate). Overhang: how far into exon do I allow read to align. Want to allow most of read to be in one exon. Usually read length minus 1. Longer = slower. 

**Step 2 - align reads**

Output to BAM = binary SAM.

Header: list of reference sequences. Usually aligner call too.
Body: One line per alignment. One read can therefore appear on multiple lines, if it aligned to multiple genes. 


SAM flag: list of binary questions. 1 = yes. See [http://broadinstitute.github.io/picard/explain-flags.html](http://broadinstitute.github.io/picard/explain-flags.html)

CIGAR string: how the read aligned in this position.

e.g.  starting at position 5, 3 matches (3M), one insertion (1I), 3M, one deletion (1D), 5M

doesn't tell about base mismatches, just if read aligned with reference. 

Do quality control with fastQC. Can filter out bad reads, but usually not done for RNA: just not aligned. Quality typically better earlier in reads. 

**Step 3 - Estimating abundances via overlap counting**

featureCounts probably most popular

Slide 33: split gene? Splicing:  introns, exons. etc. not usually seen in viruses. see slide 22

Count paired end data once only. 



### Alignment-free approach

Take all transcripts (e.g. cDNA), form all equivalence classes. Map reads to classes. Estimate underlying abundencies given observations, using EM. Only sees exons, not introns. 

Effective length: adjusting for biases and the fact that the fragment has to start a certain number of bases from the end of the fragment. Also, high gc content can mean no reads from a segment (adjsut for this).

Which transcripts correspond to which gene: tximport.

2 transcripts can be isoforms of same gene. 


c_i : numver reads, length l. 

TPM: transcript per million 


## Tutorial RNAseq

iSee: visualising SummarizedExperiment data: reproducible. 

## Bulk RNA-seq analysis

Different kinds of expression.

Always give raw counts.

### Challenges for RNA-seq data analysis- 

#### Few samples -> difficult to estimate parameters. Shrinkage: 

assume dispersion is the same for all the genes = Strong assumption. Shrinkage - assume that similarly expressed genes have similar disperson. Better estimate of dispersion of some genes.

DEseq2: fit curve (parammetric or not); move dots towards regression line.

#### Challenges for RNA-seq data analysis- High dimensionality (many genes) -> many tests

Having the transcript abundancies allows for the average transcript length to be calculated. 


# Day 5

## Single cell RNA seq

Reverse transcription very inefficient : 10-20% capturing/reverse transcription efficiency. 

Gene's express in bursting patterns i.e. not always expressing. Single cell resolution over time needed (gene turning on and off).

Note: cell identity represents the combined effect of cell type (permanent feature, e.g. neuron) and cell state (transient feature, e.g. cell cycle stage)

gel bead contains oleagone nucleotides. droplets can contain 0,1, multiple cells/beads: stochastic. 

droplet based: fewer reads per cell, only gene level resolution and not transcript level protocol. low SNR. Bursting means more zeros too. 

plate based: labour intensive, but transcript level resolution. 

droplet: good for discovering rate 

[https://www.elveflow.com/microfluidic-tutorials/microfluidic-reviews-and-tutorials/drop-seq/](https://www.elveflow.com/microfluidic-tutorials/microfluidic-reviews-and-tutorials/drop-seq/)


PCR handle: PCR handle is in fact a constant sequence identical on all primers and beads, which allows PCR amplification after STAMP formation.
UMI: unique molecular identifier. assign read to trascript of origin
barcode: assign read to cell of origin.

tSNE: only locally interprebale. small distances means close expression, but large distances don't mean anything. 

### zero-inflated negative binomial

Detection rate is +vly correlated with first pc.  


## Proteomics

[https://en.wikipedia.org/wiki/Tandem_mass_spectrometry](https://en.wikipedia.org/wiki/Tandem_mass_spectrometry)

M/Z: Mass over charge

Choose one peptide and reanalyse: break up

